### PR TITLE
fix(cgroups): PodMetadata.IPs as low level net.IP

### PR DESCRIPTION
**Pull request by nerzhul**
We are currently working on finding what happen on #33063, and during the debug logging analysis and the code reading i found that pod metadata fetch on our very busy clusters is done very very very often, hundreds of time per second, because it's done per socket creation.

Each time we retrieve those metadata we currently parse string adresses we copy over the metadata store to the caller through channels and we parse it on each metadata call.

With this change, we parse it on the store directly, once and we can directly play with it on the callers without converting each time we call it. This will reduce CPU cycles each time we create a socket, and also the memory of the metadata store because we use a raw 4 bytes instead of 15 bytes (for IPv4) string

```release-note
Reduce memory footprint for cgroup PodMetadata tracing & CPU usage on each socket creation.
```

note: it's totally backportable on 1.15 branch

### CODE REVIEW

**Review by aditighag — APPROVED** [review]
Thanks! Jfyi, we recently optimized the related code path to cache pod metadata so that channel invocations are reduced - https://github.com/cilium/cilium/pull/32615.

**Comment by nerzhul** [comment]
> Thanks! Jfyi, we recently optimized the related code path to cache pod metadata so that channel invocations are reduced - #32615.

glad to see those changes because it's a very heavy code path on our side with may explain our sync issues on pod changes or help it.
I'll fix the patch english typos tomorrow

**Comment by nerzhul** [comment]
rebased, typo fixes for the CI in the commit msg, it should be ready for a merge :)

**Review by markpash — APPROVED** [review]

**Comment by aditighag** [comment]
There've been efforts to migrate toward using `netip.addr` so I would expect we would want to continue with that trend. I've posted this in the #development cilium channel - https://cilium.slack.com/archives/C2B917YHE/p1719534771075679.

**Comment by nerzhul** [comment]
i take this as, i need to rewrite my patch to use this :) i'll do this in the upcoming days

**Comment by aditighag** [comment]
@rolinh @ldelossa Could you review the PR for your respective codeowners in the meantime?

**Comment by rolinh** [comment]
Hubble changes lgtm once we use `netip.Addr`. It'll also spare us a `net.IP` to `netip.Addr` [conversion](https://github.com/cilium/cilium/blob/7db9251cb0a523e0117e9666f8a3e3f2bb50ed41/pkg/hubble/parser/sock/parser.go#L91).

**Comment by joestringer** [comment]
@rolinh to be clear are you expecting another PR for the netip change before we merge this or does this already provide value so we could merge it?

**Review by ldelossa — APPROVED** [review]
LGTM :+1:

**Comment by nerzhul** [comment]
then, it's ok for a merge ?  change required in the future ? :)

**Review by rolinh — APPROVED** [review]

**Comment by rolinh** [comment]
> @rolinh to be clear are you expecting this PR to change for netip, or another PR for the netip change before we merge this, or does this already provide value so we could merge it?

The patch is fine as-is but moving to netip would've killed two birds with one stone. I don't want to block the PR on this however.

**Comment by joestringer** [comment]
@nerzhul There are many tests failing, so those will need to be resolved before merging.

File: pkg/cgroups/manager/manager.go
Comment (squeed) [inline pkg/cgroups/manager/manager.go:79]: Just adding a comment that this should be a `netip.Addr`. That way we won't accidentally merge.

**Comment by youngnick** [comment]
@nerzhul, looks like you're almost all good on approvals, you just need to get CI to green. I'd recommend a rebase from `main` and see if that changes anything.

**Comment by nerzhul** [comment]
i just did a naive rebase, i'll try to take time to fix CI afterwards this week

**Comment by nerzhul** [comment]
@youngnick CI is better it seems

**Comment by youngnick** [comment]
Looks like there's also one change from @squeed for you to make @nerzhul, then this seems good to go.

**Comment by joestringer** [comment]
I'll mark this PR as draft until you have addressed the outstanding comments. Once you've done that, please click the "Ready for Review" button at the bottom of the page to bring this back to our attention.

**Comment by nerzhul** [comment]
@joestringer i don't have time anymore to work on this, i changed to another company, feel free to adopt and complete it

**Comment by markpash** [comment]
> @joestringer i don't have time anymore to work on this, i changed to another company, feel free to adopt and complete it

@nerzhul I'll take it, thanks for taking it this far. 😄

**Comment by joestringer** [comment]
@nerzhul no worries, thanks for contributing! And thanks @markpash for volunteering to take this the last step :rocket:

**Pull request closed without merging**
